### PR TITLE
Fix multiple issues by switching database drivers

### PR DESCRIPTION
**Pull request by jchampio**
`lib/pq` has been in maintenance mode for a while, and issue https://github.com/timescale/timescaledb-parallel-copy/issues/61 appears to have run into one of its idiosyncrasies: its `COPY` implementation assumes that you're using a query generated via `pq.CopyIn()`, which uses the default `TEXT` format, so it runs all of the incoming data through an additional escaping layer.

Our code uses `CSV` by default (and there appears to be no way to use `TEXT` format, since we're using the old `COPY` syntax), which means that incoming CSV containing its own escapes will be double-escaped and corrupted. This is most visible with `bytea` columns, but I found similar issues with tab and backslash characters, and there are probably other problematic cases too.

To fix, switch from `lib/pq` over to `jackc/pgx`. We were already depending on a part of this library before, so the new dependency isn't as big of a change as it would have been otherwise, but the switch isn't free. The compiled binary gains roughly 1.5 MB in size -- likely due to `pgx`'s extensive type conversion system, which is unfortunate because we're not using it. Further optimization could probably be done, at the expense of having most of the DB logic go through the low-level APIs rather than `database/sql`.

I've backfilled several tests to try to characterize the previous and new behavior. The new logic has been moved to `db.CopyFromLines()` for easier testing. While I was doing that, I took a look through some of the open issues and found that #31 just happens to be fixed by this change as well, presumably because `pgx` isn't forcing a specific `DateStyle`. The new interface also makes it easy to grab the actual number of rows touched, so I went ahead and fixed our row count reporting for multi-line CSV rows. (But note that the core bug for #19 and #50 remains unfixed.)

The fact that multiple issues are fixed at once is nice, but it highlights what a big change this would be. It's probably safe to assume that there are other subtle changes in behavior. Assuming this approach is acceptable, I need to continue finding and testing corner cases to try to better characterize all of those user-visible changes. I also need to develop some real-world test cases, both for sanity checking and for performance/stress testing.

WDYT?

### Roadmap of the patchset
- move and backfill tests for the old `processBatch()` implementation, renaming it to `db.CopyFromLines()`
- switch out the `lib/pq` driver for `jackc/pgx/v4/stdlib`, updating expected test behavior
- remove dead code after the switch
- fix the reported row count using the `pgconn`'s `COPY` result
- further backfill timestamp tests now that `DateStyle` copies appear to be fixed
- **new:** move and backfill tests for `scan()`, renaming it to `batch.Scan()`
- **new:** reimplement `batch.Scan()` using `net.Buffers` and `bufio.Reader` for a considerable performance improvement

### Miscellanea
- minimum Go version has been bumped to 1.13
- utility version has been bumped to 0.4.0-dev (this is a major breaking change)
- to avoid skipping the new tests, you must specify `TEST_CONNINFO` (see the README)
- **new:** the `-token-size` option is no longer applicable and has been removed

### CODE REVIEW

File: internal/db/db.go [hunk 1]
```diff
@@ -143,9 +146,63 @@ func Connect(connStr string, overrides ...Overrideable) (*sqlx.DB, error) {
...
+	defer conn.Close()
+
+	// Our slice of lines needs to be presented to PgConn.CopyFrom() as a
+	// unified io.Reader. Joining the entire string would eat memory, so use an
+	// intermediary io.Pipe instead.
+	//
+	// TODO: just take the bytes directly as an argument rather than converting
+	// to a slice and back
+	r, w := io.Pipe()
+	go func() {
```
Comment (jchampio) [inline internal/db/db.go:172]: This pipe approach seems to be causing a significant performance hit compared to `lib/pq`, so I'll look into fixing this TODO.

**Review by svenklemm — APPROVED** [review]

**Comment by jchampio** [comment]
@svenklemm Thanks for the review! Would you prefer that I fix the new performance regression before merging?

**Comment by svenklemm** [comment]
Either is fine with me so pick whatever you prefer.

**Comment by jchampio** [comment]
I'll plan to fix this in the same PR, then. The slowdown is pretty significant and I have an approach that is looking much faster in preliminary testing.

**Comment by jchampio** [comment]
Okay, the `scan()` implementation has been moved to its own package for testing, and I've rewritten it to pull `[]byte` slices directly into the `io.Reader` used by `PgConn`. This leads to an overall performance improvement over the `lib/pq` implementation, and the CPU usage of the utility drops considerably since we're not copying as much data around.

**Review by svenklemm — APPROVED** [review]

**Comment by jchampio** [comment]
Thanks for the reviews!
